### PR TITLE
WIP system + actors + workers

### PR DESCRIPTION
**Pull request by twokul**
### TL;DR about Event based vs. Thread base actor systems

Languages (like Scala) make the distinction between `thread-based` and `event-based` actors.

Thread-based actors are actors which each run in their own JVM thread. They are scheduled by the Java thread scheduler, which uses a preemptive priority-based scheduler (when the actor enters a `receive` block, the thread is blocked until messages arrive). Thread-based actors make it possible to do long-running computations, or blocking I/O operations inside actors, without hindering the execution of other actors.

There is an important drawback to this method: each thread can be considered as being heavy-weight and uses a certain amount of memory and imposes some scheduling overhead.

Event-based actors are not on the same thread. An actor that waits for a message to be received is not represented by a blocked thread, but by a closure. This closure captures the state of the actor, such that it’s computation can be continued upon receiving a message. The execution of this closure happens on the thread of the sender. There's [a cool paper](http://lampwww.epfl.ch/~odersky/papers/jmlc06.pdf) about it.

Event-based actors provide a more light-weight alternative, allowing for very large numbers of concurrently running actors. They should however not be used for parallelism: since all actors execute on the same thread, there is no scheduling fairness.

### API

```javascript
var Ara = require('ara');
var System = Ara.System;
var Actor = Ara.Actor;

var BabelActor = Actor.extend({
  receive: function(value) {
    var result = require('babel-core').transform(value);

    return {
      code: result.code,
      map: result.map
    };
  }
});

// snippet from `broccoli-babel-transpiler`
Babel.prototype.processString = function(string, relativePath) {
  // ... code ...
  var babelActor = system.actorOf(BabelActor);

  return babelActor.send({
    string: string,
    options: options
  }).then(function(transpiled) {
    // ... code ...
    return transpiled.code;
  });
};
```

Side note: we should likely change `send` to `ask` to be more consistent with the rest of actor world.

The experiment above allows you to create unlimited number of actors in the system but only use `maxCPU` processes. Everything returns a promise as of now (meaning, there's an option to `ask` and to `tell` actor to do something if you will).

Creation of actors on their own is not supported, aka `let actor = new Actor();`, I doubt we need it but I might be wrong.

Process pool is also lazy (doesn't do work unless you explicitly ask it to). Scheduling is pretty stupid ATM, needs to be re-visited after we get first benchmarks. This is basically a **WIP** to get @stefanpenner feedback whenever he has time.

I will also try to attach before/after benchmarks of our [`stress-app`](https://github.com/ember-cli/stress-app).

### integration with ember-cli

the only change I had to make it work with ember-cli was to patch `broccoli-babel-transpiler` (I assume this needs more exploring).

```diff
diff --git a/index.js b/index.js
index 0866b32..5bfa572 100644
--- a/index.js
+++ b/index.js
@@ -1,5 +1,6 @@
 'use strict';
 
+var Ara        = require('ara');
 var transpiler = require('babel-core');
 var Filter     = require('broccoli-persistent-filter');
 var clone      = require('clone');
@@ -11,6 +12,22 @@ var funnel     = require('broccoli-funnel');
 var crypto     = require('crypto');
 var hashForDep = require('hash-for-dep');
 
+var System = Ara.System;
+var Actor = Ara.Actor;
+
+var BabelActor = Actor.extend({
+  receive: function(value) {
+    var result = require('babel-core').transform(value);
+
+    return {
+      code: result.code,
+      map: result.map
+    };
+  }
+});
+
+var system = new System();
+
 function getExtensionsRegex(extensions) {
   return extensions.map(function(extension) {
     return new RegExp('\.' + extension + '$');
@@ -164,6 +181,7 @@ Babel.prototype.cacheKeyProcessString = function(string, relativePath) {
 
 Babel.prototype.processString = function(string, relativePath) {
   var options = this.copyOptions();
+  var helperWhiteList = this.helperWhiteList;
 
   options.filename = options.sourceMapTarget = options.sourceFileName = relativePath;
 
@@ -171,17 +189,34 @@ Babel.prototype.processString = function(string, relativePath) {
     options.moduleId = replaceExtensions(this.extensionsRegex, options.filename);
   }
 
-  var transpiled = this.transform(string, options);
+  var babelActor = system.actorOf(BabelActor);
+  return babelActor.send({
+    string: string,
+    options: options
+  }).then(function(transpiled) {
+    if (helperWhiteList) {
+      var invalidHelpers = transpiled.metadata.usedHelpers.filter(function(helper) {
+        return helperWhiteList.indexOf(helper) === -1;
+      });
+
+      validateHelpers(invalidHelpers, relativePath);
+    }
 
-  if (this.helperWhiteList) {
-    var invalidHelpers = transpiled.metadata.usedHelpers.filter(function(helper) {
-      return this.helperWhiteList.indexOf(helper) === -1;
-    }, this);
+    console.log('broccoli-babel-transpiler: code', transpiled.code);
+    return transpiled.code;
+  });
 
-    validateHelpers(invalidHelpers, relativePath);
-  }
+  //var transpiled = this.transform(string, options);
+
+  //if (this.helperWhiteList) {
+  //  var invalidHelpers = transpiled.metadata.usedHelpers.filter(function(helper) {
+  //    return this.helperWhiteList.indexOf(helper) === -1;
+  //  }, this);
+
+  //  validateHelpers(invalidHelpers, relativePath);
+  //}
 
-  return transpiled.code;
+  //return transpiled.code;
 };
 
 Babel.prototype.copyOptions = function() {
```

### CODE REVIEW

File: dist/ara.js [hunk 1]
```diff
@@ -0,0 +1,464 @@
+'use strict';
```
Comment (twokul) [inline dist/ara.js]: remove this stuff from here

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (twokul) [inline .vscode/launch.json:1]: remove this stuff from here

File: lib/actor.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    this.id = uuid.v4();
+  }
+
+  async send(message) {
+    let receive = this.receive;
+    /*
+      direct all messages to parent (system) as it should know
+      how to schedule work using process pool & workers.
+    */
+    return this.parent.schedule({ message, receive });
```
Comment (twokul) [inline lib/actor.js]: so this seems wrong and needs to be replaced with `send` to the guardian actor

File: lib/process-pool.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+      this.workers[lastIndex]
+    ] = [
+      this.workers[lastIndex],
+      this.workers[firstIndex]
+    ];
+
+    return result;
+  }
+
+  popIdleWorker() {
```
Comment (twokul) [inline lib/process-pool.js]: this function needs to be renamed as it doesn't just returns a worker but does weird iffy "scheduling"; maybe replace it with priority scheduling? or queue based one, backburner.js-ish; "IDLE", "SCHEDULED"

File: lib/system.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+/* eslint-enable */
+
+const debug = _debug('ara:system');
+
+export default class System {
+  constructor() {
+    this.processPool = new ProcessPool();
+  }
+
+  async schedule() {
```
Comment (twokul) [inline lib/system.js]: this looks very simplistic, do we need to do anything else here?

File: lib/process-pool.js [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        console.log('ara:schedule - first run new worker');
+        this.workers.push(worker);
+        this.idle.push(worker);
+        resolve(worker);
+      }
+
+      if (this.isIdlePoolEmpty()) {
+        // this means all workers are busy
+        console.log('ara:schedule - busy');
+        setTimeout(() => {
```
Comment (stefanpenner) [inline lib/process-pool.js]: rather then spinning on idle, it may make more sense to store an queue (just an array) of work, so that when workers become idle they can pop up that queue.

File: lib/process-pool.js [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    //   this._becameIdle(worker);
+    //   return value;
+    // } catch(reason) {
+    //   this._crashed(worker, reason);
+    //   throw reason;
+    // }
+  // }
+
+  _becameIdle(worker) {
+    this.idle.push(worker); // ensure unique, maybe use fast-ordered-set
```
Comment (stefanpenner) [inline lib/process-pool.js]: the other comment, about puling a queue of pending work and processing on idle. Likely the hook to process idle, could happen here. This way we don't need to spin on idle, but rather something going idle automatically checks for any queue'd up work.

File: lib/process-pool.js [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    ];
+
+    return result;
+  }
+
+  popIdleWorker() {
+    return new Promise((resolve, reject) => {
+      if (this.isWorkersPoolEmpty()) {
+        let worker = new Worker(`${__dirname}/runner.js`);
+        console.log('ara:schedule - first run new worker');
```
Comment (stefanpenner) [inline lib/process-pool.js]: rather then using `console.log` could we use `heimdalljs-logger` ? https://github.com/heimdalljs/heimdalljs-logger

File: lib/process-pool.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+        console.log('ara:schedule - busy');
+        setTimeout(() => {
+          return resolve(this.popIdleWorker());
+        }, 1);
+      } else {
+        let worker;
+
+        if (this.isMaxWorkers()) {
+          console.log('ara:schedule - max workers');
+          worker = this.popWorker();
```
Comment (stefanpenner) [inline lib/process-pool.js]: it seems like we may want to push into a pending work queue here instead, then let a future idle worker take the work.

File: lib/process-pool.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    //   throw reason;
+    // }
+  // }
+
+  _becameIdle(worker) {
+    this.idle.push(worker); // ensure unique, maybe use fast-ordered-set
+  }
+
+  _crashed(worker, reason) {
+    // remove from workers
```
Comment (stefanpenner) [inline lib/process-pool.js]: I suspect we need to discuss strategies for recovering or failing on failure.

some useful strategies:

1. if one worker fails, the whole `Pool` fails. The work is not disjoint, and the pool cannot proceed
2. if one worker fails, kill the failed worker and start a new one. The work is disjoint, and the pool can continue on.

For build tools, most likely the first option makes the most sense. As once a task fails, its game over anyways.

File: lib/system.js (hunk 5 above)
Comment (stefanpenner) [reply lib/system.js]: It's fine. I suspect if we want more functionality, it will be my specifying a specific type of priority queue. But this API most likely wont need to do much.

The only thing we should consider, is the ability to cancel work. But we can do that in a later roud.

**Review by stefanpenner — CHANGES_REQUESTED** [review]

**Comment by stefanpenner** [comment]
> Event-based actors provide a more light-weight alternative, allowing for very large numbers of concurrently running actors. They should however not be used for parallelism: since all actors execute on the same thread, there is no scheduling fairness.

it seems like an event based actor can run both in the same thread, or in another. Which allows for it both as a unit of parallelism and as a unit for concurrency. Although parallelism is optional aspect is optional. Seems good for our usecase :)

File: lib/actor.js [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+import CoreObject from 'core-object';
+import uuid from 'node-uuid';
+/* eslint-disable */
+import regeneratorRuntime from 'regenerator-runtime';
+/* eslint-enable */
+import _debug from 'debug';
+
+const debug = _debug('actor');
+
+export default class Actor extends CoreObject {
```
Comment (stefanpenner) [inline lib/actor.js]: i would prefer if we dropped core-object.

**Review by stefanpenner — CHANGES_REQUESTED** [review]

File: lib/process-pool.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+import Worker from './worker';
+import './ensure-promise';
+/* eslint-disable */
+import regeneratorRuntime from 'regenerator-runtime';
+/* eslint-enable */
+
+// TODO: this file is just a brainstorm sketch, it actually needs to be
+// implemented.
+
+const maxCPU = 4;
```
Comment (stefanpenner) [inline lib/process-pool.js]: can we use the existing maxCPU module?

File: lib/process-pool.js [hunk 13]
```diff
@@ -0,0 +1,131 @@
+// import maxCPU from './max-cpu';
```
Comment (stefanpenner) [inline lib/process-pool.js]: do you mind leaving this as `pool` that way the diff carriers over.

File: lib/process-pool.js [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+// implemented.
+
+const maxCPU = 4;
+
+export default class Pool {
+  constructor(options) {
+    // this.j = options && options.j || maxCPU;
+    this.workers = [];
+    this.idle = [];
+    // this.inbox = [];
```
Comment (stefanpenner) [inline lib/process-pool.js]: this is most likely where our pending work goes, to then later be processed.

File: lib/process-pool.js [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  popIdleWorker() {
+    return new Promise((resolve, reject) => {
+      if (this.isWorkersPoolEmpty()) {
+        let worker = new Worker(`${__dirname}/runner.js`);
+        console.log('ara:schedule - first run new worker');
+        this.workers.push(worker);
+        this.idle.push(worker);
+        resolve(worker);
```
Comment (stefanpenner) [inline lib/process-pool.js]: the work is now `idle`, but when is it removed from idle? `popIdleWorker` should likely not add the worker to idle.

File: lib/process-pool.js [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    ] = [
+      this.workers[lastIndex],
+      this.workers[firstIndex]
+    ];
+
+    return result;
+  }
+
+  popIdleWorker() {
+    return new Promise((resolve, reject) => {
```
Comment (stefanpenner) [inline lib/process-pool.js]: can we use async/await here, like the rest?

File: lib/process-pool.js [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+      this._becameIdle(worker);
+      return value;
+    } catch(reason) {
+      this._crashed(worker, reason);
+      throw reason;
+    }
+  }
+
+  async ask(arg) {
+    // this.inbox.push(arg);
```
Comment (stefanpenner) [inline lib/process-pool.js]: I think we should go back to pushing onto the inbox, and letting the next available worker pull off that queue.

When scheduling, we should make sure the max number of workers are alive and otherwise create one more.

**Review by stefanpenner — CHANGES_REQUESTED** [review]

File: lib/process-pool.js (hunk 8 above)
Comment (twokul) [reply lib/process-pool.js]: we absolutely should
